File: core/transcription/src/commonMain/kotlin/dev/audiocompanion/transcription/OpenAiTranscriptionProvider.kt
package dev.audiocompanion.transcription

import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.forms.MultiPartFormDataContent
import io.ktor.client.request.forms.formData
import io.ktor.client.request.header
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.http.Headers
import io.ktor.http.HttpHeaders
import io.ktor.http.HttpStatusCode
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.io.Buffer
import kotlinx.io.readByteArray
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlin.math.roundToLong

/**
 * Cloud transcription provider for bounded, durable segment chunks.
 *
 * Uses OpenAI's Audio API `transcriptions` endpoint. PCM is uploaded only when the current
 * transcription mode can use cloud transcription and an API key is configured.
 */
class OpenAiTranscriptionProvider(
    private val client: HttpClient,
    private val apiKey: () -> String?,
    private val cloudConsent: () -> Boolean,
    private val model: () -> String = { DEFAULT_MODEL },
    /**
     * When true, transcription is routed through the diarization model
     * ([DIARIZE_MODEL], `diarized_json`) so segments carry a `speaker`. This overrides [model],
     * since only the diarize model returns speaker labels. The tradeoff is no word-level timings.
     */
    private val diarizationEnabled: () -> Boolean = { false },
    private val endpointUrl: String = DEFAULT_ENDPOINT_URL,
    private val maxUploadBytes: Int = DEFAULT_MAX_UPLOAD_BYTES,
) : TranscriptionProvider, CloudUploadCapable {
    override val id: String = "openai"
    override val status: StateFlow<ProviderStatus> = MutableStateFlow(ProviderStatus.Ready)

    private val json = Json { ignoreUnknownKeys = true }

    init {
        require(maxUploadBytes > WAV_HEADER_BYTES) {
            "maxUploadBytes must leave room for a WAV header"
        }
    }

    override suspend fun isAvailable(): Boolean =
        cloudConsent() && !apiKey().isNullOrBlank()

    override suspend fun transcribe(
        pcmChunks: Flow<ByteArray>,
        sampleRateHz: Int,
    ): TranscriptionResult {
        if (!cloudConsent()) {
            throw TranscriptionException.ProviderUnavailable(id)
        }
        val key = apiKey()?.takeIf { it.isNotBlank() }
            ?: throw TranscriptionException.ProviderUnavailable(id)

        val transcripts = mutableListOf<OpenAiChunkResult>()
        var chunkIndex = 0
        var chunkStartMs = 0L
        val pcmUploadLimit = maxUploadBytes - WAV_HEADER_BYTES
        val buffer = Buffer()
        try {
            pcmChunks.collect { chunk ->
                var offset = 0
                while (offset < chunk.size) {
                    val writable = minOf(pcmUploadLimit - buffer.size.toInt(), chunk.size - offset)
                    if (writable <= 0) {
                        val pcm = buffer.readByteArray(buffer.size.toInt())
                        transcripts += uploadChunk(
                            key = key,
                            pcm = pcm,
                            sampleRateHz = sampleRateHz,
                            chunkIndex = chunkIndex++,
                            chunkStartMs = chunkStartMs,
                        )
                        chunkStartMs += pcm.durationMs(sampleRateHz)
                        continue
                    }
                    buffer.write(chunk, offset, offset + writable)
                    offset += writable
                }
            }
            if (buffer.size > 0) {
                val pcm = buffer.readByteArray(buffer.size.toInt())
                transcripts += uploadChunk(
                    key = key,
                    pcm = pcm,
                    sampleRateHz = sampleRateHz,
                    chunkIndex = chunkIndex,
                    chunkStartMs = chunkStartMs,
                )
            }
        } finally {
            buffer.close()
        }

        val text = transcripts.joinToString(separator = "\n") { it.text }.trim()
        if (text.isBlank()) {
            throw TranscriptionException.NoSpeechDetected("OpenAI returned an empty transcript")
        }
        return TranscriptionResult(
            text = text,
            providerId = id,
            modelUsed = effectiveModel(),
            segments = transcripts.flatMap { it.segments },
            words = transcripts.flatMap { it.words },
        )
    }

    /** The diarize model takes precedence when diarization is on; only it returns speakers. */
    private fun effectiveModel(): String = if (diarizationEnabled()) DIARIZE_MODEL else model()

    private fun responseFormat(model: String): String = when (model) {
        DIARIZE_MODEL -> "diarized_json"
        "whisper-1" -> "verbose_json"
        else -> "json"
    }

    private suspend fun uploadChunk(
        key: String,
        pcm: ByteArray,
        sampleRateHz: Int,
        chunkIndex: Int,
        chunkStartMs: Long,
    ): OpenAiChunkResult {
        if (pcm.isEmpty()) {
            throw TranscriptionException.NoSpeechDetected("empty PCM chunk")
        }
        val wav = PcmWav.encodeMono16(pcm, sampleRateHz)
        if (wav.size > maxUploadBytes) {
            throw TranscriptionException.TranscriptionFailed(
                "WAV upload chunk ${wav.size} exceeds $maxUploadBytes byte limit",
            )
        }
        val modelValue = effectiveModel()
        val format = responseFormat(modelValue)
        val response: HttpResponse = client.post(endpointUrl) {
            header(HttpHeaders.Authorization, "Bearer $key")
            setBody(
                MultiPartFormDataContent(
                    formData {
                        append("model", modelValue)
                        append("response_format", format)
                        if (format == "verbose_json") {
                            append("timestamp_granularities[]", "segment")
                            append("timestamp_granularities[]", "word")
                        }
                        append(
                            "file",
                            wav,
                            Headers.build {
                                append(HttpHeaders.ContentType, "audio/wav")
                                append(
                                    HttpHeaders.ContentDisposition,
                                    "filename=\"segment-$chunkIndex.wav\"",
                                )
                            },
                        )
                    },
                ),
            )
        }
        val body = response.body<String>()
        if (response.status != HttpStatusCode.OK) {
            throw TranscriptionException.TranscriptionFailed(
                "OpenAI transcription failed (${response.status.value}): ${body.take(240)}",
            )
        }
        return parseChunk(body, chunkStartMs)
    }

    private fun parseChunk(body: String, offsetMs: Long): OpenAiChunkResult =
        when (responseFormat(effectiveModel())) {
            "verbose_json" ->
                json.decodeFromString(OpenAiVerboseTranscriptionResponse.serializer(), body)
                    .toChunkResult(offsetMs)
            "diarized_json" ->
                json.decodeFromString(OpenAiDiarizedTranscriptionResponse.serializer(), body)
                    .toChunkResult(offsetMs)
            else ->
                OpenAiChunkResult(
                    text = json.decodeFromString(OpenAiTranscriptionResponse.serializer(), body).text,
                )
        }

    // --- CloudUploadCapable: single-shot background upload (response is the transcript) ----------

    override suspend fun uploadPlan(wav: ByteArray, sampleRateHz: Int): CloudUploadPlan? {
        if (!cloudConsent()) return null
        val key = apiKey()?.takeIf { it.isNotBlank() } ?: return null
        // One background upload is a single request; segments larger than the API limit stay on the
        // synchronous chunked path.
        if (wav.size > maxUploadBytes) return null
        val modelValue = effectiveModel()
        val format = responseFormat(modelValue)
        val fields = buildList {
            add("model" to modelValue)
            add("response_format" to format)
            if (format == "verbose_json") {
                add("timestamp_granularities[]" to "segment")
                add("timestamp_granularities[]" to "word")
            }
        }
        return CloudUploadPlan(
            url = endpointUrl,
            headers = mapOf("Authorization" to "Bearer $key"),
            textFields = fields,
            file = MultipartBody.FilePart("file", "segment.wav", "audio/wav", wav),
        )
    }

    override suspend fun onUploadResponse(httpStatus: Int, body: String): CloudUploadStep {
        if (httpStatus !in 200..299) {
            throw TranscriptionException.TranscriptionFailed(
                "OpenAI transcription failed ($httpStatus): ${body.take(240)}",
            )
        }
        val chunk = parseChunk(body, 0L)
        val text = chunk.text.trim()
        if (text.isBlank()) {
            throw TranscriptionException.NoSpeechDetected("OpenAI returned an empty transcript")
        }
        return CloudUploadStep.Done(
            TranscriptionResult(
                text = text,
                providerId = id,
                modelUsed = effectiveModel(),
                segments = chunk.segments,
                words = chunk.words,
            ),
        )
    }

    override suspend fun completeControlPlane(controlState: String): TranscriptionResult =
        throw TranscriptionException.TranscriptionFailed("OpenAI uploads have no control-plane step")

    @Serializable
    private data class OpenAiTranscriptionResponse(val text: String = "")

    @Serializable
    private data class OpenAiVerboseTranscriptionResponse(
        val text: String = "",
        val segments: List<OpenAiTimedText> = emptyList(),
        val words: List<OpenAiTimedWord> = emptyList(),
    ) {
        fun toChunkResult(offsetMs: Long): OpenAiChunkResult =
            OpenAiChunkResult(
                text = text,
                segments = segments.mapNotNull { segment ->
                    val cleaned = segment.text.trim()
                    if (cleaned.isBlank()) return@mapNotNull null
                    TranscriptSegment(
                        text = cleaned,
                        startMs = offsetMs + (segment.start * 1_000).roundToLong(),
                        endMs = offsetMs + (segment.end * 1_000).roundToLong(),
                    )
                },
                words = words.mapNotNull { word ->
                    val cleaned = word.word.trim()
                    if (cleaned.isBlank()) return@mapNotNull null
                    TranscriptWord(
                        text = cleaned,
                        startMs = offsetMs + (word.start * 1_000).roundToLong(),
                        endMs = offsetMs + (word.end * 1_000).roundToLong(),
                    )
                },
            )
    }

    @Serializable
    private data class OpenAiDiarizedTranscriptionResponse(
        val text: String = "",
        val segments: List<OpenAiDiarizedSegment> = emptyList(),
    ) {
        fun toChunkResult(offsetMs: Long): OpenAiChunkResult =
            OpenAiChunkResult(
                text = text,
                segments = segments.mapNotNull { segment ->
                    val cleaned = segment.text.trim()
                    if (cleaned.isBlank()) return@mapNotNull null
                    TranscriptSegment(
                        text = cleaned,
                        startMs = offsetMs + (segment.start * 1_000).roundToLong(),
                        endMs = offsetMs + (segment.end * 1_000).roundToLong(),
                        speaker = segment.speaker,
                    )
                },
            )
    }

    @Serializable
    private data class OpenAiDiarizedSegment(
        val text: String = "",
        val start: Double = 0.0,
        val end: Double = 0.0,
        val speaker: String? = null,
    )

    @Serializable
    private data class OpenAiTimedText(
        val text: String = "",
        val start: Double = 0.0,
        val end: Double = 0.0,
    )

    @Serializable
    private data class OpenAiTimedWord(
        val word: String = "",
        val start: Double = 0.0,
        val end: Double = 0.0,
    )

    private data class OpenAiChunkResult(
        val text: String,
        val segments: List<TranscriptSegment> = emptyList(),
        val words: List<TranscriptWord> = emptyList(),
    )

    private fun ByteArray.durationMs(sampleRateHz: Int): Long =
        (size / BYTES_PER_SAMPLE.toDouble() / sampleRateHz * 1_000).roundToLong()

    companion object {
        const val DEFAULT_ENDPOINT_URL = "https://api.openai.com/v1/audio/transcriptions"
        const val DEFAULT_MODEL = "gpt-4o-mini-transcribe"

        /** OpenAI's speaker-diarization model; returns `diarized_json` with per-segment speakers. */
        const val DIARIZE_MODEL = "gpt-4o-transcribe-diarize"
        private const val WAV_HEADER_BYTES = 44
        private const val BYTES_PER_SAMPLE = 2
        private const val DEFAULT_MAX_UPLOAD_BYTES = 24_000_000
    }
}

File: ios/PebbleAudioKit/Sources/Transcription/Providers/OpenAiTranscriptionProvider.swift
import AudioCodec
import Foundation

// Port of `core/transcription/.../OpenAiTranscriptionProvider.kt`.

/// Cloud transcription provider for bounded, durable segment chunks.
///
/// Uses OpenAI's Audio API `transcriptions` endpoint. PCM is uploaded only when the current
/// transcription mode can use cloud transcription and an API key is configured. Keys arrive via
/// the injected closure (Keychain lookups live in the app layer — never read from disk here).
public final class OpenAiTranscriptionProvider: TranscriptionProvider, CloudUploadCapable {
    public static let defaultEndpointUrl = "https://api.openai.com/v1/audio/transcriptions"
    public static let defaultModelsUrl = "https://api.openai.com/v1/models"
    public static let defaultModel = "gpt-transcribe"

    /// OpenAI's speaker-diarization model; returns `diarized_json` with per-segment speakers.
    public static let diarizeModel = "gpt-4o-transcribe-diarize"

    static let wavHeaderBytes = 44
    public static let defaultMaxUploadBytes = 24_000_000
    private static let bytesPerSample = 2

    public let id = "openai"

    private let transport: any HttpTransport
    private let apiKey: @Sendable () -> String?
    private let cloudConsent: @Sendable () -> Bool
    private let model: @Sendable () -> String
    /// When true, transcription is routed through the diarization model (`diarizeModel`,
    /// `diarized_json`) so segments carry a `speaker`. This overrides `model`, since only the
    /// diarize model returns speaker labels. The tradeoff is no word-level timings.
    private let diarizationEnabled: @Sendable () -> Bool
    /// Bounded keyword list for STT steering; ignored for diarize models.
    private let sttPrompt: @Sendable () -> String?
    let endpointUrl: String
    let modelsUrl: String
    private let maxUploadBytes: Int

    public init(
        transport: any HttpTransport,
        apiKey: @escaping @Sendable () -> String?,
        cloudConsent: @escaping @Sendable () -> Bool,
        model: @escaping @Sendable () -> String = { OpenAiTranscriptionProvider.defaultModel },
        diarizationEnabled: @escaping @Sendable () -> Bool = { false },
        sttPrompt: @escaping @Sendable () -> String? = { nil },
        endpointUrl: String = OpenAiTranscriptionProvider.defaultEndpointUrl,
        modelsUrl: String = OpenAiTranscriptionProvider.defaultModelsUrl,
        maxUploadBytes: Int = OpenAiTranscriptionProvider.defaultMaxUploadBytes
    ) {
        precondition(
            maxUploadBytes > Self.wavHeaderBytes,
            "maxUploadBytes must leave room for a WAV header"
        )
        self.transport = transport
        self.apiKey = apiKey
        self.cloudConsent = cloudConsent
        self.model = model
        self.diarizationEnabled = diarizationEnabled
        self.sttPrompt = sttPrompt
        self.endpointUrl = endpointUrl
        self.modelsUrl = modelsUrl
        self.maxUploadBytes = maxUploadBytes
    }

    public func isAvailable() async -> Bool {
        cloudConsent() && !(apiKey() ?? "").isBlank
    }

    // Connectivity-probe seam (see CloudConnectivityChecks.swift).
    var connectivityTransport: any HttpTransport { transport }
    var connectivityApiKey: String? { apiKey() }

    public func transcribe(
        pcmChunks: AsyncThrowingStream<Data, Error>,
        sampleRateHz: Int
    ) async throws -> TranscriptionResult {
        guard cloudConsent() else {
            throw TranscriptionError.providerUnavailable(providerId: id)
        }
        guard let key = apiKey()?.nonBlank else {
            throw TranscriptionError.providerUnavailable(providerId: id)
        }

        var transcripts: [ChunkResult] = []
        var chunkIndex = 0
        var chunkStartMs: Int64 = 0
        let pcmUploadLimit = maxUploadBytes - Self.wavHeaderBytes
        var buffer = Data()

        for try await chunk in pcmChunks {
            var offset = 0
            while offset < chunk.count {
                let writable = min(pcmUploadLimit - buffer.count, chunk.count - offset)
                if writable <= 0 {
                    let pcm = buffer
                    buffer = Data()
                    transcripts.append(
                        try await uploadChunk(
                            key: key,
                            pcm: pcm,
                            sampleRateHz: sampleRateHz,
                            chunkIndex: chunkIndex,
                            chunkStartMs: chunkStartMs
                        )
                    )
                    chunkIndex += 1
                    chunkStartMs += Self.durationMs(byteCount: pcm.count, sampleRateHz: sampleRateHz)
                    continue
                }
                let start = chunk.startIndex + offset
                buffer.append(chunk[start..<(start + writable)])
                offset += writable
            }
        }
        if !buffer.isEmpty {
            transcripts.append(
                try await uploadChunk(
                    key: key,
                    pcm: buffer,
                    sampleRateHz: sampleRateHz,
                    chunkIndex: chunkIndex,
                    chunkStartMs: chunkStartMs
                )
            )
        }

        let text = transcripts.map(\.text).joined(separator: "\n").trimmed
        if text.isEmpty {
            throw TranscriptionError.noSpeechDetected("OpenAI returned an empty transcript")
        }
        return TranscriptionResult(
            text: text,
            providerId: id,
            modelUsed: effectiveModel(),
            segments: transcripts.flatMap(\.segments),
            words: transcripts.flatMap(\.words)
        )
    }

    /// The diarize model takes precedence when diarization is on; only it returns speakers.
    func effectiveModel() -> String {
        diarizationEnabled() ? Self.diarizeModel : model()
    }

    private func sttPrompt(forModel modelValue: String) -> String? {
        if modelValue.lowercased().contains("diarize") { return nil }
        return sttPrompt()?.trimmed.nonBlank
    }

    private func responseFormat(model: String) -> String {
        switch model {
        case Self.diarizeModel: return "diarized_json"
        case "whisper-1": return "verbose_json"
        default: return "json"
        }
    }

    private func uploadChunk(
        key: String,
        pcm: Data,
        sampleRateHz: Int,
        chunkIndex: Int,
        chunkStartMs: Int64
    ) async throws -> ChunkResult {
        if pcm.isEmpty {
            throw TranscriptionError.noSpeechDetected("empty PCM chunk")
        }
        let wav = PcmWav.encodeMono16(pcm: pcm, sampleRateHz: sampleRateHz)
        if wav.count > maxUploadBytes {
            throw TranscriptionError.transcriptionFailed(
                "WAV upload chunk \(wav.count) exceeds \(maxUploadBytes) byte limit"
            )
        }
        let (fields, filePart) = requestParts(wav: wav, filename: "segment-\(chunkIndex).wav")
        let boundary = "PebbleAudioKit-\(UUID().uuidString)"
        let (body, contentType) = MultipartBody.encode(
            boundary: boundary, textFields: fields, file: filePart
        )
        let response = try await transport.execute(
            HttpTransportRequest(
                method: "POST",
                url: endpointUrl,
                headers: [
                    "Authorization": "Bearer \(key)",
                    "Content-Type": contentType,
                ],
                body: body
            )
        )
        if response.status != 200 {
            throw TranscriptionError.transcriptionFailed(
                "OpenAI transcription failed (\(response.status)): \(response.text.prefix(240))"
            )
        }
        return try parseChunk(body: response.body, offsetMs: chunkStartMs)
    }

    /// The multipart text fields + file part shared by the synchronous path and `uploadPlan`.
    private func requestParts(
        wav: Data,
        filename: String
    ) -> (fields: [(String, String)], file: MultipartBody.FilePart) {
        let modelValue = effectiveModel()
        let format = responseFormat(model: modelValue)
        var fields: [(String, String)] = [
            ("model", modelValue),
            ("response_format", format),
        ]
        if format == "verbose_json" {
            fields.append(("timestamp_granularities[]", "segment"))
            fields.append(("timestamp_granularities[]", "word"))
        }
        if let prompt = sttPrompt(forModel: modelValue) {
            fields.append(("prompt", prompt))
        }
        let file = MultipartBody.FilePart(
            name: "file", filename: filename, contentType: "audio/wav", bytes: wav
        )
        return (fields, file)
    }

    private func parseChunk(body: Data, offsetMs: Int64) throws -> ChunkResult {
        let decoder = JSONDecoder()
        switch responseFormat(model: effectiveModel()) {
        case "verbose_json":
            return try decoder.decode(VerboseResponse.self, from: body).toChunkResult(offsetMs: offsetMs)
        case "diarized_json":
            return try decoder.decode(DiarizedResponse.self, from: body).toChunkResult(offsetMs: offsetMs)
        default:
            let plain = try decoder.decode(PlainResponse.self, from: body)
            return ChunkResult(text: plain.text ?? "")
        }
    }

    // MARK: - CloudUploadCapable: single-shot background upload (the response IS the transcript)

    public func uploadPlan(wav: Data, sampleRateHz: Int) async -> CloudUploadPlan? {
        guard cloudConsent(), let key = apiKey()?.nonBlank else { return nil }
        // One background upload is a single request; segments larger than the API limit stay on
        // the synchronous chunked path.
        if wav.count > maxUploadBytes { return nil }
        let (fields, filePart) = requestParts(wav: wav, filename: "segment.wav")
        return CloudUploadPlan(
            url: endpointUrl,
            headers: ["Authorization": "Bearer \(key)"],
            textFields: fields,
            file: filePart
        )
    }

    public func onUploadResponse(httpStatus: Int, body: String) async throws -> CloudUploadStep {
        guard (200..<300).contains(httpStatus) else {
            throw TranscriptionError.transcriptionFailed(
                "OpenAI transcription failed (\(httpStatus)): \(body.prefix(240))"
            )
        }
        let chunk = try parseChunk(body: Data(body.utf8), offsetMs: 0)
        let text = chunk.text.trimmed
        if text.isEmpty {
            throw TranscriptionError.noSpeechDetected("OpenAI returned an empty transcript")
        }
        return .done(
            TranscriptionResult(
                text: text,
                providerId: id,
                modelUsed: effectiveModel(),
                segments: chunk.segments,
                words: chunk.words
            )
        )
    }

    public func completeControlPlane(controlState: String) async throws -> TranscriptionResult {
        throw TranscriptionError.transcriptionFailed("OpenAI uploads have no control-plane step")
    }

    // MARK: - Response shapes

    private struct PlainResponse: Decodable {
        var text: String?
    }

    private struct VerboseResponse: Decodable {
        var text: String?
        var segments: [TimedText]?
        var words: [TimedWord]?

        func toChunkResult(offsetMs: Int64) -> ChunkResult {
            ChunkResult(
                text: text ?? "",
                segments: (segments ?? []).compactMap { segment in
                    let cleaned = (segment.text ?? "").trimmed
                    if cleaned.isEmpty { return nil }
                    return TranscriptSegment(
                        text: cleaned,
                        startMs: offsetMs + msFromSeconds(segment.start ?? 0),
                        endMs: offsetMs + msFromSeconds(segment.end ?? 0)
                    )
                },
                words: (words ?? []).compactMap { word in
                    let cleaned = (word.word ?? "").trimmed
                    if cleaned.isEmpty { return nil }
                    return TranscriptWord(
                        text: cleaned,
                        startMs: offsetMs + msFromSeconds(word.start ?? 0),
                        endMs: offsetMs + msFromSeconds(word.end ?? 0)
                    )
                }
            )
        }
    }

    private struct DiarizedResponse: Decodable {
        var text: String?
        var segments: [DiarizedSegment]?

        func toChunkResult(offsetMs: Int64) -> ChunkResult {
            ChunkResult(
                text: text ?? "",
                segments: (segments ?? []).compactMap { segment in
                    let cleaned = (segment.text ?? "").trimmed
                    if cleaned.isEmpty { return nil }
                    return TranscriptSegment(
                        text: cleaned,
                        startMs: offsetMs + msFromSeconds(segment.start ?? 0),
                        endMs: offsetMs + msFromSeconds(segment.end ?? 0),
                        speaker: segment.speaker
                    )
                }
            )
        }
    }

    private struct DiarizedSegment: Decodable {
        var text: String?
        var start: Double?
        var end: Double?
        var speaker: String?
    }

    private struct TimedText: Decodable {
        var text: String?
        var start: Double?
        var end: Double?
    }

    private struct TimedWord: Decodable {
        var word: String?
        var start: Double?
        var end: Double?
    }

    private struct ChunkResult {
        var text: String
        var segments: [TranscriptSegment] = []
        var words: [TranscriptWord] = []
    }

    private static func durationMs(byteCount: Int, sampleRateHz: Int) -> Int64 {
        Int64(
            (Double(byteCount) / Double(bytesPerSample) / Double(sampleRateHz) * 1_000)
                .rounded(.toNearestOrAwayFromZero)
        )
    }
}

/// Seconds-to-milliseconds with Kotlin `roundToLong` semantics (round-half-up for the positive
/// timestamps these APIs produce).
private func msFromSeconds(_ seconds: Double) -> Int64 {
    Int64((seconds * 1_000).rounded(.toNearestOrAwayFromZero))
}

// Kotlin isBlank/isNotBlank analogues (fileprivate to stay collision-free within the module).
extension String {
    fileprivate var trimmed: String { trimmingCharacters(in: .whitespacesAndNewlines) }
    fileprivate var isBlank: Bool { trimmed.isEmpty }
    /// Self when non-blank, else nil (`takeIf { it.isNotBlank() }`).
    fileprivate var nonBlank: String? { isBlank ? nil : self }
}
